interfaces/builtin/greengrass_support.go

```diff
@@ -86,38 +94,134 @@ capability dac_read_search,
 capability sys_admin,
 capability dac_override,  # for various overlayfs accesses
 
+# for setting up mounts
 @{PROC}/[0-9]*/mountinfo r,
 @{PROC}/filesystems r,
 
-# setup the overlay so we may pivot_root into it
+# cgroup accesses
+# greengrassd extensively uses cgroups to confine it's containers (AKA lambdas)
+# and needs to read what cgroups are available; we allow reading any cgroup, 
+# but limit writes below
+# also note that currently greengrass is not implemented in such a way that it
+# can stack it's cgroups inside the cgroup that snapd would normally enforce
+# but this may change in the future
+# an example cgroup access looks like this:
+# /old_rootfs/sys/fs/cgroup/cpuset/system.slice/7d23e67f-13f5-4b7e-5a85-83f8773345a8/
+# the old_rootfs prefix is due to the pivot_root - the "old" rootfs is mounted
+# at /old_rootfs before
+@{PROC}/cgroups r,
+owner @{PROC}/[0-9]*/cgroup r,
+owner /old_rootfs/sys/fs/cgroup/{,**} r,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,system.slice/}system.slice/ rw,
+owner /old_rootfs/sys/fs/cgroup/{blkio,cpuset,devices,hugetlb,memory,perf_event,pids,freezer/snap.greengrass}/{,system.slice/}system.slice/[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f]-[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f]/{,**} rw,
```

Comment:
`/{,system.slice/}system.slice` evaluates to /system.slice or system.slice/system.slice-- is that intended? Is it this way because of the snap.greengrass alternative? If so, I wonder if you can have a couple snap.greengrass rules and a couple non-snap.greengrass rules instead of both on the same line. (I like to use the {a,b} alternates when it makes things still easy to read/audit.

You are also allowing writes to the system blkio,..., was that intentional?

The thoughts behind these questions apply to the next set of cgroup rules too. Please consider if they could be written more clearly while being somewhat concise.

Reply:
So this is basically the full set of accesses I see for first Ubuntu 18.04 classic:

```
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.cpus
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.mems
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

and on Ubuntu 16.04 classic:

```
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.cpus
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/cpuset.mems
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/pids/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

and on Ubuntu Core 16:
```
/old_rootfs/sys/fs/cgroup/blkio/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/blkio/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/cpu,cpuacct/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/cpuset/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/devices/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/freezer/snap.greengrass/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/hugetlb/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/memory/system.slice/system.slice/
/old_rootfs/sys/fs/cgroup/memory/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/net_cls,net_prio/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/perf_event/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/pids/system.slice/system.slice/$UUID
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/
/old_rootfs/sys/fs/cgroup/systemd/system.slice/snap.greengrass.greengrassd.service/system.slice/$UUID
```

I do not know why some of the cgroups are for `system.slice/snap.greengrass.greengrassd.service` on classic 18.04 and some are for `system.slice/system.slice` on 16.04 versions, the only thing I can think of is different kernel and thus different cgroup versions? Or perhaps different systemd versions?

Reply:
Weird. Those uuid rules are fairly terrible. AppArmor or snapd could add a @{uuid} variable to make these easier to read. That isn't for this PR.

These rules are fine.